manifests/profile/quod/dev/perl.pp

```diff
@@ -0,0 +1,90 @@
+# Copyright (c) 2020 The Regents of the University of Michigan.
+# All Rights Reserved. Licensed according to the terms of the Revised
+# BSD License. See LICENSE.txt for details.
+
+# nebula::profile::quod::dev::perl
+#
+# Install perl dependencies for quod applications. These packages
+# and settings are unique to the dev environment.
+#
+# @example
+#   include nebula::profile::quod::dev::perl
+class nebula::profile::quod::dev::perl () {
+  require nebula::profile::quod::dependencies::perl
+
+  package { [
+    'libany-moose-perl',
+    'libapt-pkg-perl',
+    'libclass-data-accessor-perl',
+    'libclass-data-inheritable-perl',
+    'libclass-errorhandler-perl',
+    'libclass-singleton-perl',
+    'libclone-perl',
+    'libconfig-file-perl',
+    'libconfig-tiny-perl',
+    'libcrypt-openssl-bignum-perl',
+    'libcrypt-openssl-random-perl',
+    'libcrypt-openssl-rsa-perl',
+    'libdatetime-locale-perl',
+    'libdatetime-perl',
+    'libdatetime-timezone-perl',
+    'libdevel-overloadinfo-perl',
+    'libdigest-hmac-perl',
+    'libemail-messageid-perl',
+    'libemail-mime-contenttype-perl',
+    'libemail-mime-encodings-perl',
+    'libemail-mime-perl',
+    'libemail-valid-perl',
+    'libexception-class-perl',
+    'libexporter-tiny-perl',
+    'libfile-basedir-perl',
+    'libfile-fnmatch-perl',
+    'libfile-homedir-perl',
+    'libfile-next-perl',
+    'libfile-stripnondeterminism-perl',
+    'libfilesys-df-perl',
+    'libfile-which-perl',
+    'libio-socket-inet6-perl',
+    'libipc-shareable-perl',
+    'libipc-system-simple-perl',
+    'liblist-allutils-perl',
+    'liblist-someutils-perl',
+    'liblist-utilsby-perl',
+    'liblog-dispatch-perl',
+    'liblog-log4perl-perl',
+    'libmouse-perl',
+    'libnamespace-autoclean-perl',
+    'libnet-dns-perl',
+    'libnet-idn-encode-perl',
+    'libnet-ip-perl',
+    'libossp-uuid-perl',
+    'libparams-validate-perl',
+    'libparams-validationcompiler-perl',
+    'libparse-debianchangelog-perl',
+    'libreadonly-perl',
+    'libreadonly-xs-perl',
+    'libregexp-assemble-perl',
+    'libscalar-list-utils-perl',
+    'libsocket6-perl',
+    'libsort-naturally-perl',
+    'libspecio-perl',
+    'libtest-fatal-perl',
+    'libtext-levenshtein-perl',
+    'libtie-ixhash-perl',
+    'libtypes-serialiser-perl',
+    'libuniversal-require-perl',
+    'libwww-curl-perl',
+    'libxml-twig-perl',
+    'libxml-xpathengine-perl',
+    'libxml-xpath-perl',
+    'libyaml-appconfig-perl',
+    'libyaml-libyaml-perl',
+    'xmlformat-perl']:
+  }
+
+  -> nebula::cpan { [
```

Comment:
Do these CPAN resources really depend on every single one of these packages?

Reply:
Yeah, I guess I might clarify into two questions...

1. Do we know why we are using the debs for these, given that we have a CPAN resource type?
2. Do we need the dependency/order relationship for this to work?

I would hazard that the first is a bit of "it worked before" and the second is some collision between system/package-tracked dependencies and CPAN-tracked dependencies unless we order the installs... but it would be good to sort this out.

It is a _lot_ of cpan-as-deb packages, and I'm wondering if we could do them all as `nebula::cpan` or use [Carton](https://metacpan.org/pod/Carton) to normalize them.

Reply:
I am basing this off of www_lib format to some degree.

Also, I am not sure which of the 200+ packages are prerequisites for the CPAN modules.

Is there an issue if these are ordered? Cause finding chains of dependencies can quite tedious and time consuming without much benefit normally.

Reply:
Part of the reasoning at least is "it works now as is so let's duplicate what is and not what it could be." Currently we should probably not change this and refactor later on I think if we feel it is important.

And I think @aelkiss or @skorner may have a better idea why we have some as apt packages and some as cpan and whether that can or should be changed. Or at least visiting at this point or in the future.

I think it is a design decision honestly. Error on deb packages for stability and a primary point of management and only the exceptions should be CPAN. 

Carton could be something to consider. But is that adding unnecessary complexity or do we gain from it is something to visit also in the future.

Reply:
I guess I'm interested to see what happens if we remove the `->`. I'm completely ok with some packages requiring other packages, but 3 CPAN installs, each requiring 67 apt packages, is a little over the top. For example, what is the worst thing that could happen? Is it that puppet attempts to run these out of order, some succeed while others fail, but then they all succeed 30 minutes later at the next sync?

Or are we worried about something breaking more seriously? What I'd like to avoid is the paranoid copying and pasting of configurations that none of us understand, since all it leads to is the continued copying and pasting of configurations that we may or may not need.

In other words, if we need that `->`, I would like for us to understand why we need it.

Reply:
Our general principle was to install perl packages from Debian if they were available and use CPAN only if there are not upstream Debian packages.

If some of the CPAN packages were installed before some of the apt packages, then they will automatically bring in their dependencies as CPAN packages; whereas if we have the apt packages installed first they will use the apt packages. For www-lib and HathiTrust I did not particularly want to keep track of the dependencies for perl modules, I just want to ensure that all the Debian-provided packages are installed before the CPAN ones.

While we are using the system perl modules -- rather than going to all-CPAN I would actually argue that it might be better to use `dh-make-perl` to create debian packages that are installed in our local Debian archive. At the least, that way we have more consistent control over versioning -- if we go all-CPAN, then as it is now you will just get whatever random version happens to be released when we first set up a machine and then they'll never be upgraded.

I agree carton would be nice, and I think it's likely I'll look into that for the HT apps that we aren't immediately migrating off perl.

Reply:
I see; that makes sense. In general I'm in favor of choosing one or the other, but in the meantime, perhaps a single line

```puppet
Package<| |> -> Nebula::Cpan<| |>
```

somewhere that applies to every system with perl could stand for all?

Reply:
I've just added an ADR entry to the PR to capture the discussion above. As Aaron has mentioned elsewhere, this is a long-standing preference for OS-packaged versions where available. The preferred model may be quite different in containers but, for system packages, this is our best option.

I have no position on the Puppet ordering mechanics / syntax.

Reply:
@daaang, do you consider this and the following comments resolved?

Reply:
I do like the ordering default given that likely that is our model for approaching things until we start making our own debs always. 

It may be a good thing to do next sprint after we get the upgrades done next week.

I would like to do it as I haven't done something like that before.

Reply:
Sounds good.